### PR TITLE
Move everything from processorhelper to component.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Updates https://github.com/open-telemetry/opentelemetry-collector/issues/4681

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: component/processor.go [hunk 1]
```diff
@@ -102,3 +103,108 @@ type ProcessorFactory interface {
...
 	) (LogsProcessor, error)
 }
+
+// ProcessorFactoryOption apply changes to ProcessorOptions.
+type ProcessorFactoryOption func(o *processorFactory)
+
+// ProcessorCreateDefaultConfigFunc is the equivalent of ProcessorFactory.CreateDefaultConfig()
+type ProcessorCreateDefaultConfigFunc func() config.Processor
+
+// CreateDefaultConfig implements ExtensionFactory.CreateDefaultConfig()
```
Comment (dmitryax) [inline component/processor.go:113]:
```suggestion
// CreateDefaultConfig implements ProcessorFactory.CreateDefaultConfig()
```

File: component/factories_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+				assert.Error(t, err)
+				return
+			}
+			assert.NoError(t, err)
+			assert.Equal(t, tt.out, out)
+		})
+	}
+}
+
+func TestMakeExtensionFactoryMap(t *testing.T) {
```
Comment (dmitryax) [inline component/factories_test.go:63]: Looks like unrelated test. Should it go in another PR?

File: component/processor.go [hunk 3]
```diff
@@ -54,8 +55,8 @@ type ProcessorCreateSettings struct {
 	BuildInfo BuildInfo
 }
 
-// ProcessorFactory is factory interface for processors. This is the
-// new factory type that can create new style processors.
+// ProcessorFactory is processorFactory interface for processors. This is the
+// new processorFactory type that can create new style processors.
 //
 // This interface cannot be directly implemented. Implementations must
 // use the processorhelper.NewFactory to implement it.
```
Comment (dmitryax) [inline component/processor.go:62]:
```suggestion
// use the NewProcessorFactory to implement it.
```

File: component/processor.go (hunk 1 above)
Comment (bogdandrutu) [reply component/processor.go:113]: Done.

File: component/processor.go (hunk 3 above)
Comment (bogdandrutu) [reply component/processor.go:62]: Done.

File: component/factories_test.go (hunk 2 above)
Comment (bogdandrutu) [reply component/factories_test.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/4904

**Comment by bogdandrutu** [comment]
@dmitryax PTAL.

File: component/factories_test.go (hunk 2 above)
Comment (dmitryax) [reply component/factories_test.go]: Looks like another test was moved to that PR :)

File: component/factories_test.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+func TestMakeProcessorFactoryMap(t *testing.T) {
```
Comment (bogdandrutu) [inline component/factories_test.go:25]: Moved this test from `processor_test.go` since it tests functionality in `factories.go`

**Review by dmitryax — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Based on the fact that multiple approvers already approved similar changes to this for component helpers and extension helpers, I will merge this.
